File: Hudson River Trading/245.shortest-word-distance-iii.py
"""
Approach: Single scan tracking recent relevant positions.
Data structure: integer indexes track the latest occurrence of each target word.
Interview logic: if the words differ, update the matching latest index and compare when both exist. If the words are the same, the shortest distance is between consecutive occurrences of that word.
Complexity: O(n) time, O(1) space.
Tests and edge cases: word1 == word2; adjacent occurrences produce distance 1; non-adjacent repeated targets update the best distance.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def shortestWordDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        best = float('inf')
        if word1 == word2:
            prev = -1
            for i, word in enumerate(wordsDict):
                if word == word1:
                    if prev != -1:
                        best = min(best, i - prev)
                    prev = i
            return best
        last1 = last2 = -1
        for i, word in enumerate(wordsDict):
            if word == word1:
                last1 = i
                if last2 != -1:
                    best = min(best, abs(last1 - last2))
            elif word == word2:
                last2 = i
                if last1 != -1:
                    best = min(best, abs(last1 - last2))
        return best
# @lc code=end

File: Palantir/245.shortest-word-distance-iii.go
package leetcode

// ShortestWordDistance245 handles equal words by tracking the previous occurrence
// of that same word. For different words, it tracks the latest index of each word
// and updates the best distance whenever the other has been seen.
//
// Time: O(n). Space: O(1).
func ShortestWordDistance245(wordsDict []string, word1 string, word2 string) int {
	best := 1<<31 - 1
	if word1 == word2 {
		prev := -1
		for i, w := range wordsDict {
			if w == word1 {
				if prev != -1 {
					best = minInt(best, i-prev)
				}
				prev = i
			}
		}
		return best
	}
	last1, last2 := -1, -1
	for i, w := range wordsDict {
		if w == word1 {
			last1 = i
			if last2 != -1 {
				best = minInt(best, absInt(last1-last2))
			}
		} else if w == word2 {
			last2 = i
			if last1 != -1 {
				best = minInt(best, absInt(last1-last2))
			}
		}
	}
	return best
}
